math/check_prime.cpp

```diff
@@ -2,55 +2,85 @@
+ * https://cp-algorithms.com/algebra/primality_tests.html#practice-problems
  */
 
 #include <cassert>   /// for assert
 #include <iostream>  /// for IO operations
 
 /**
- * Function to check if the given number is prime or not.
+ * @brief Function to check if the given number is prime or not.
  * @param num number to be checked.
- * @return if number is prime, it returns @ true, else it returns @ false.
+ * @return true if number is a prime
+ * @return false if number is not a prime.
  */
-template <typename T>
-bool is_prime(T num) {
-    bool result = true;
+bool is_prime(long long num) {
+    /*!
+     * Reduce all possibilities of a number which cannot be prime with the first
+     * 3 if, else if conditionals. Example: Since no even number, except 2 can
+     * be a prime number and the next prime we find after our checks is 5,
+     * we will start the for loop with i = 5. then for each loop we increment
+     * i by +6 and check if i or i+2 is a factor of the number; if it's a factor
+     * then we will return false. otherwise, true will be returned after the
+     * loop terminates at the terminating condition which is i*i <= num
+     */
     if (num <= 1) {
         return false;
     } else if (num == 2 || num == 3) {
         return true;
-    } else if ((num % 2) == 0 || num % 3 == 0) {
+    } else if (num % 2 == 0 || num % 3 == 0) {
         return false;
     } else {
-        for (T i = 5; (i * i) <= (num); i = (i + 6)) {
-            if ((num % i) == 0 || (num % (i + 2) == 0)) {
-                result = false;
-                break;
+        for (long long i = 5; i * i <= num; i = i + 6) {
+            if (num % i == 0 || num % (i + 2) == 0) {
+                return false;
             }
         }
     }
-    return (result);
+    return true;
 }
 
 /**
- * Main function
+ * @brief Self-test implementations
+ * @returns void
  */
-int main() {
-    // perform self-test
+static void tests() {
+    /*!
+     * each assert test checks if the is_prime function returns the expected
+     * boolean value to whether its argument is prime
+     */
+    assert(is_prime(1) == false);
+    assert(is_prime(2) == true);
+    assert(is_prime(3) == true);
+    assert(is_prime(4) == false);
+    assert(is_prime(-4) == false);
+    assert(is_prime(7) == true);
+    assert(is_prime(-7) == false);
+    assert(is_prime(19) == true);
     assert(is_prime(50) == false);
     assert(is_prime(115249) == true);
 
-    int num = 0;
+    std::cout << "All tests have successfully passed!" << std::endl;
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    tests();  // perform self-tests implementations
+    long long num = 0;
     std::cout << "Enter the number to check if it is prime or not" << std::endl;
-    std::cin >> num;
+    std::cin >> num;  // take user chosen number and store in variable num
     bool result = is_prime(num);
     if (result) {
         std::cout << num << " is a prime number" << std::endl;
```

Comment:
imo we should remove using std::cin/cout in main as it clutters execution and i would rather it be left as just for tests.
but thats just me

Reply:
I guess it depends on how you view it, I like a little bit of interactivity, helps make things more rememberable even if its as simple as this, especially for new programmers. However, I think for consistency in the repo we will remove it.